java_api/file_api/src/main/java/uk/ramp/metadata/MetadataItem.java

```diff
@@ -100,57 +108,94 @@ default MetadataItem applyOverrides(List<OverrideItem> overrides) {
 
     MetadataItem overriddenMetadataItem = ImmutableMetadataItem.copyOf(this);
     for (MetadataItem override : overridesToApply) {
-      overriddenMetadataItem = applyOverride(overriddenMetadataItem, override);
+      overriddenMetadataItem = applyOverride(overriddenMetadataItem, override, onlyOverrideIfEmpty);
     }
     return overriddenMetadataItem;
   }
 
-  private MetadataItem applyOverride(MetadataItem baseMetadata, MetadataItem metadataOverride) {
+  private boolean shouldOverrideField(
+      Function<MetadataItem, Boolean> isFieldPresent,
+      MetadataItem baseMetadata,
+      MetadataItem overrideMetadata,
+      boolean onlyOverrideIfEmpty) {
+    boolean isBaseFieldPresent = isFieldPresent.apply(baseMetadata);
+    boolean isOverrideFieldPresent = isFieldPresent.apply(overrideMetadata);
+
+    if (!isOverrideFieldPresent) {
+      return false;
+    }
+    if (onlyOverrideIfEmpty) {
+      return !isBaseFieldPresent;
+    }
+    return true;
+  }
+
+  private MetadataItem applyOverride(
+      MetadataItem baseMetadata, MetadataItem metadataOverride, boolean onlyOverrideIfEmpty) {
     var newMetadataItem = ImmutableMetadataItem.copyOf(baseMetadata);
 
-    if (metadataOverride.internalFilename().isPresent()) {
+    Function<Function<MetadataItem, Boolean>, Boolean> shouldOverride =
+        fieldFunc ->
+            shouldOverrideField(fieldFunc, baseMetadata, metadataOverride, onlyOverrideIfEmpty);
+
+    Function<MetadataItem, Boolean> isFilenamePresent = meta -> meta.internalFilename().isPresent();
+    Function<MetadataItem, Boolean> isComponentPresent = meta -> meta.component().isPresent();
+    Function<MetadataItem, Boolean> isDataProductPresent = meta -> meta.dataProduct().isPresent();
+    Function<MetadataItem, Boolean> isInternalVersionPresent =
+        meta -> meta.internalVersion().isPresent();
+    Function<MetadataItem, Boolean> isExtensionPresent = meta -> meta.extension().isPresent();
+    Function<MetadataItem, Boolean> isVerifiedHashPresent = meta -> meta.verifiedHash().isPresent();
+    Function<MetadataItem, Boolean> isCalculatedHashPresent =
+        meta -> meta.calculatedHash().isPresent();
+    Function<MetadataItem, Boolean> isRunIdPresent = meta -> meta.runId().isPresent();
```

Comment:
@mrow84 - is run id still a thing?

Reply:
Yes, it is used to generate unique filenames on write.